polarCoordinateTraversal.m

```diff
@@ -111,75 +111,134 @@
     end
     return;
 end
 
 % If there is a ray/shell intersection, then set the radial voxel ID. 
 current_voxel_ID_r = 1 + (circle_max_radius - r)/delta_radius;
 if verbose
-    fprintf('RADIAL HIT INITIALIZED.\n');
+    fprintf('RADIAL HIT INITIALIZED.\n')
 end
 
-% Calculate Voxel ID Theta.
-current_voxel_ID_theta = floor(atan2(ray_start_y - circle_center_y, ray_start_x - circle_center_x) * num_angular_sections / (2 * pi));
+% II. Calculate Voxel ID Theta.
+tol = 10^-16;
+if abs(ray_origin - circle_center) < tol 
+    % If the ray starts at the origin, we need to perturb slightly along its path to find the
+    % correct angular voxel
+    pert_t = 0.1;
+    pert_x = ray_start_x + ray_direction_x * pert_t;
+    pert_y = ray_start_y + ray_direction_y * pert_t;
+    current_voxel_ID_theta = floor(atan2(pert_y - circle_center_y, pert_x - circle_center_x) * num_angular_sections / (2 * pi));
+else
+    current_voxel_ID_theta = floor(atan2(ray_start_y - circle_center_y, ray_start_x - circle_center_x) * num_angular_sections / (2 * pi));
+end
 if current_voxel_ID_theta < 0
     current_voxel_ID_theta = num_angular_sections + current_voxel_ID_theta;
 end
 
-% Pre-calculations for radial_hit. This is necessary to check for intersections with relevant radial neighbors.
-ray_unit_vector = 1 / sqrt(ray_direction(1)^2 + ray_direction(2)^2)...
-    .* [ray_direction(1);  ray_direction(2)]';
-ray_circle_vector = [circle_center(1) - ray_origin(1); circle_center(2) - ray_origin(2)]';
-v = dot(ray_circle_vector,ray_unit_vector);
-
-
 angular_voxels = [current_voxel_ID_theta];
 radial_voxels = [current_voxel_ID_r];
 
-% TRAVERSAL PHASE
-t = t_begin; 
-previous_transition_flag = false;
+% Find the maximum time the ray will be in the grid
+discr = circle_max_radius^2 - (dot(ray_circle_vector,ray_circle_vector) - v^2);
+d = sqrt(discr);
+pa_max = ray_origin + (v-d).*ray_unit_vector;
+pb_max = ray_origin + (v+d).*ray_unit_vector;
+if ray_direction(1) < tol
+    t1 = (pa_max(2) - ray_origin(2))/ray_direction(2);
+    t2 = (pb_max(2) - ray_origin(2))/ray_direction(2);
+else
+    t1 = (pa_max(1) - ray_origin(1))/ray_direction(1);
+    t2 = (pb_max(1) - ray_origin(1))/ray_direction(1);
+end
+t_grid = max(t1,t2);
 
-while (t < t_end)
+% III. TRAVERSAL PHASE
+t = t_begin;
+previous_transition_flag = false;
+while t < min(t_grid,t_end)
+    
     % 1. Calculate tMaxR, tMaxTheta
     [tMaxR, tStepR, previous_transition_flag] = radial_hit(ray_origin, ray_direction, ...
         current_voxel_ID_r, circle_center, circle_max_radius, delta_radius, t, ray_unit_vector, ray_circle_vector, v, previous_transition_flag, verbose);
     [tMaxTheta, tStepTheta] = angular_hit(ray_origin, ray_direction, current_voxel_ID_theta,...
         num_angular_sections, circle_center, t, verbose);
     
     % 2. Compare tMaxTheta, tMaxR
-    if (tMaxTheta < tMaxR)
+    if tMaxTheta < tMaxR || current_voxel_ID_r + tStepR == 0 && t < tMaxTheta && tMaxTheta < min(t_grid,t_end)
```

Comment:
Can we use parentheses here? I don't remember priority for ```&&``` versus```||```, but from my understanding, this code is saying that if the next radial step exits our circle, we should do an angular step, which, correct me if I'm wrong, is not always the case.

Reply:
"if the next radial step exits our circle, we should do an angular step" IF the angular step can still be taken within the grid time.

Reply:
Order of ops. should be left to right, so parens won't matter: chain of conjunctions

Reply:
"if the next radial step exits our circle, we should do an angular step" IF the angular step can still be taken within the grid time. ---- Before, we were exiting the grid automatically dependent on tStepR without checking if we could still take an angular step.

Reply:
According to the [documentation](https://www.mathworks.com/help/matlab/matlab_prog/operator-precedence.html), there is a precedence order for MATLAB as well.

And ok this makes sense.